### ISSUE TITLE
Hash is incorrect in safari,if pass path using unicode string to navigate method 

### ISSUE THREAD

**Issue by leexiaosi** [jashkenas/backbone#3857]
In Safari , pass a path using unicode string to `navigate` method  like this

``` js
Backbone.history.navigate('/city/北京',{trigger:true});
```

The hash is incorrect. It output

``` js
location.hash
//> #/city/%17%AC
```

I tracked the code, it caused by `_updateHash` method.

The call stack created by  `History.prototype.navigate` method like this
- navigate
  -  decodeFragment
  - _updateHash

The fragment passed to `_updateHash`,  was decoded by `decodeFragment`, so the hash was incorrect。

I also searched the commit history, the `decodeFragment` added in "1.2.0"

PS: My english is not good, although I am trying to improve

**Comment by jridgewell** [comment]
Seems to be working for me. Can you provide a jsfiddle showing the error?

**Comment by leexiaosi** [comment]
[https://jsfiddle.net/leexiaosi/uyp3o6ek/](https://jsfiddle.net/leexiaosi/uyp3o6ek/)

**notice!!!** : please open in safari.

### the record of safari
![safari](https://cloud.githubusercontent.com/assets/516562/11164567/62f9ddf6-8b2f-11e5-8aa8-631c1aa84422.gif)

### the record of chrome
![chrome](https://cloud.githubusercontent.com/assets/516562/11164583/c66d1b82-8b2f-11e5-9d88-ecba16ae6167.gif)

**Comment by paulfalgout** [comment]
This seems fairly similar to what is described here: https://github.com/jashkenas/backbone/issues/3830

**Comment by rrott** [comment]
Looks like Safari are buggy in the implementation of assignment to window.location.hash =(
北京 in unicode should be %u53**17**%u4E**AC** but not **%17%AC**

I faced the same issue with Ukrainian chars as well:
>[Log] cityName is : їє (index.html, line 35)
>[Log] locationHash is : #city/WT (index.html, line 36)

It affects all non-latin chars

addinig encodeURI() in _updateHash() before assigning a new hash fixes the issue in Safari, but in all other browsers links will look like 
/index.html#city/%E5%8C%97%E4%BA%AC

>location.hash = '#' + encodeURI(fragment);
https://github.com/jashkenas/backbone/blob/master/backbone.js#L1835

I am not sure if it's ok to send such pull requests =(

**Comment by rrott** [comment]
Found a better solution. Will update you once tests are done and green

**Comment by okovpashko** [comment]
I faced with the same issue. 
Only in Safari. Sample: http://jsbin.com/rexefe/edit?html,js,console

Looks like the [test](https://github.com/jashkenas/backbone/blob/master/test/router.js#L194) for unicode symbols in the hash has already been written. It passes in Safari. 
But after some research, I figured out that window.location is [emulated](https://github.com/jashkenas/backbone/blob/master/test/router.js#L19) in tests by the `<a>` HTML element and assigning not URL encoded hash to it is worked correctly.

**Comment by Flamefork** [comment]
AFAIU, as the hash is a part of URI, unicode characters should be encoded with `encodeURIComponent`. Assigning an unicode string to `location.hash` should probably be avoided, and this Safari quirk is just another reason for this.

https://jsfiddle.net/1pht15aa/ is the original authors jsfiddle but with the latest `master` and using `encodeURLComponent`. Hash is encoded, name is consistent, both in Safari and chrome.

**Comment by jgonggrijp** [comment]
This seems to have been fixed in #4026. In case anyone runs into it anyway, please feel free to comment (or reopen if you can).
